org.eclipse.dsp4j.jsonrpc/src/test/java/org/eclipse/dsp4j/jsonrpc/test/json/DebugMessageJsonHandlerTest.java

```diff
@@ -0,0 +1,582 @@
+		Assert.assertEquals("bar", result.getRight().get(0).getValue());
+	}
+
+	@SuppressWarnings({ "unchecked" })
+	@Test
+	public void testEither_05() {
+		Map<String, JsonRpcMethod> supportedMethods = new LinkedHashMap<>();
+		supportedMethods.put("foo", JsonRpcMethod.request("foo",
+				new TypeToken<Either<List<MyClass>, MyClassList>>() {}.getType(),
+				new TypeToken<Object>() {}.getType()));
+		DebugMessageJsonHandler handler = new DebugMessageJsonHandler(supportedMethods);
+		handler.setMethodProvider((id) -> "foo");
+
+		Message message = handler.parseMessage("{"
+				+ "\"seq\":2,\n"
+				+ "\"type\":\"response\",\n"
+				+ "\"body\": [{\n"
+				+ "  value:\"foo\"\n"
+				+ "}]}");
+		Either<List<MyClass>, MyClassList> result = (Either<List<MyClass>, MyClassList>) ((ResponseMessage)message).getResult();
+		Assert.assertTrue(result.isLeft());
+		Assert.assertEquals("foo", result.getLeft().get(0).getValue());
+
+		message = handler.parseMessage("{"
+				+ "\"seq\":2,\n"
+				+ "\"type\":\"response\",\n"
+				+ "\"body\": {\n"
+				+ "  items: [{\n"
+				+ "    value:\"bar\"\n"
+				+ "}]}}");
+		result = (Either<List<MyClass>, MyClassList>) ((ResponseMessage)message).getResult();
+		Assert.assertTrue(result.isRight());
+		Assert.assertEquals("bar", result.getRight().getItems().get(0).getValue());
+	}
+
+	@Test
+	public void testParamsParsing_01() {
+		Map<String, JsonRpcMethod> supportedMethods = new LinkedHashMap<>();
+		supportedMethods.put("foo", JsonRpcMethod.request("foo",
+				new TypeToken<Void>() {}.getType(),
+				new TypeToken<Location>() {}.getType()));
+		DebugMessageJsonHandler handler = new DebugMessageJsonHandler(supportedMethods);
+		handler.setMethodProvider((id) -> "foo");
+		RequestMessage message = (RequestMessage) handler.parseMessage("{"
+				+ "\"seq\":2,\n"
+				+ "\"type\":\"request\",\n"
+				+ "\"command\":\"foo\",\n"
+				+ "\"arguments\": {\"uri\": \"dummy://mymodel.mydsl\"}\n"
+				+ "}");
+		Object params = message.getParams();
+		Class<? extends Object> class1 = params.getClass();
+		Assert.assertEquals(Location.class, class1);
+	}
+
+	public static final <T> void swap(T[] a, int i, int j) {
+		T t = a[i];
+		a[i] = a[j];
+		a[j] = t;
+	}
+
+	void testAllPermutations(String[] properties, Consumer<String> test) {
+		testAllPermutations(properties, 0, properties.length - 1, test);
+	}
+
+	void testAllPermutations(String[] properties, int i, int n, Consumer<String> test) {
+		int j;
+		if (i == n) {
+			StringBuilder json = new StringBuilder();
+			json.append("{");
+			for (int k = 0; k < properties.length; k++) {
+				json.append(properties[k]);
+				if (k != properties.length - 1) {
+					json.append(",");
+				}
+
+			}
+			json.append("}");
+			test.accept(json.toString());
+		} else {
+			for (j = i; j <= n; j++) {
+				swap(properties, i, j);
+				testAllPermutations(properties, i + 1, n, test);
+				swap(properties, i, j);
+			}
+		}
+	}
+
+	@Test
+	public void testParamsParsing_AllOrders() {
```

Comment:
This parsing test that tries all iterations would also be nice for the `MessageJsonHandlerTest` in `org.eclipse.lsp4j.jsonrpc`.

Reply:
OK, I will contribute it back to both places. I also realised on revisiting that the idea was to remove the _02/_03/etc versions of the methods, so I will do that too.

Reply:
I have added the `_AllOrders` method to `MessageJsonHandlerTest` and extended it to do all the types of messages.